File: python/src/com/jetbrains/python/debugger/containerview/PyDataViewDialog.java
/*
 * Copyright 2000-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.python.debugger.containerview;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.util.NlsContexts;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.util.ui.JBUI;
import com.jetbrains.python.PyBundle;
import com.jetbrains.python.debugger.PyDebugValue;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class PyDataViewDialog extends DialogWrapper {
  private final JPanel myMainPanel;
  private static final int TABLE_DEFAULT_WIDTH = 700;
  private static final int TABLE_DEFAULT_HEIGHT = 500;
  private final Project myProject;
  private final PyDataViewerPanel myDataViewerPanel;

  public PyDataViewDialog(@NotNull Project project, @NotNull final PyDebugValue value) {
    super(project, false);
    myProject = project;
    setModal(false);
    setCancelButtonText(PyBundle.message("debugger.data.view.close"));
    setCrossClosesWindow(true);

    myMainPanel = new JPanel(new GridBagLayout());
    myDataViewerPanel = new PyDataViewerPanel(project, value.getFrameAccessor());
    myDataViewerPanel.apply(value, false);
    myDataViewerPanel.setPreferredSize(JBUI.size(TABLE_DEFAULT_WIDTH, TABLE_DEFAULT_HEIGHT));

    myMainPanel.add(myDataViewerPanel, createDataViewPanelConstraints());

    myDataViewerPanel.addListener(new PyDataViewerPanel.Listener() {
      @Override
      public void onNameChanged(@NlsContexts.TabTitle @NotNull String name) {
        setTitle(name);
      }
    });

    addBottomElements();

    setTitle(value.getFullName());
    init();
  }

  protected void addBottomElements() {
    final JBCheckBox colored = new JBCheckBox(PyBundle.message("debugger.data.view.colored.cells"));
    final JBCheckBox resize = new JBCheckBox(PyBundle.message("debugger.data.view.resize.automatically"));

    resize.setSelected(PyDataView.Companion.isAutoResizeEnabled(myProject));
    colored.setSelected(PyDataView.Companion.isColoringEnabled(myProject));

    colored.addActionListener(new ActionListener() {
      @Override
      public void actionPerformed(ActionEvent e) {
        myDataViewerPanel.setColored(colored.isSelected());
        myDataViewerPanel.updateUI();
      }
    });
    resize.addActionListener(new ActionListener() {
      @Override
      public void actionPerformed(ActionEvent e) {
        myDataViewerPanel.resize(resize.isSelected());
        myDataViewerPanel.updateUI();
      }
    });

    GridBagConstraints checkBoxConstraints = createCheckBoxConstraints();
    checkBoxConstraints.gridy = 1;
    myMainPanel.add(colored, checkBoxConstraints);
    checkBoxConstraints.gridy = 2;
    myMainPanel.add(resize, checkBoxConstraints);
  }

  @Override
  protected Action @NotNull [] createActions() {
    return new Action[]{getCancelAction()};
  }

  @Override
  protected JComponent createCenterPanel() {
    return myMainPanel;
  }

  protected GridBagConstraints createDataViewPanelConstraints() {
    GridBagConstraints c = new GridBagConstraints();
    c.fill = GridBagConstraints.BOTH;
    c.weightx = 0.9;
    c.weighty = 0.9;
    return c;
  }

  protected GridBagConstraints createCheckBoxConstraints() {
    GridBagConstraints c = new GridBagConstraints();
    c.weighty = 0.05;
    c.anchor = GridBagConstraints.LINE_START;
    return c;
  }
}

File: python/src/com/jetbrains/python/debugger/containerview/PyDataViewDialog.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.jetbrains.python.debugger.containerview

import com.intellij.ide.IdeBundle
import com.intellij.ide.plugins.PluginManagerConfigurable
import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.updateSettings.impl.pluginsAdvertisement.installAndEnable
import com.intellij.openapi.util.NlsContexts
import com.intellij.openapi.util.NlsSafe
import com.intellij.ui.EditorNotificationPanel
import com.intellij.ui.components.JBCheckBox
import com.intellij.util.PlatformUtils
import com.intellij.util.ui.JBUI
import com.jetbrains.python.PyBundle
import com.jetbrains.python.debugger.PyDebugValue
import com.jetbrains.python.debugger.containerview.PyDataView.Companion.isAutoResizeEnabled
import com.jetbrains.python.debugger.containerview.PyDataView.Companion.isColoringEnabled
import com.jetbrains.python.debugger.containerview.PyDataView.Companion.setAutoResizeEnabled
import com.jetbrains.python.debugger.containerview.PyDataView.Companion.setColoringEnabled
import java.awt.BorderLayout
import javax.swing.Action
import javax.swing.BoxLayout
import javax.swing.JPanel

class PyDataViewDialog(private val myProject: Project, value: PyDebugValue) : DialogWrapper(myProject, false) {
  private val mainPanel: JPanel
  private val dataViewerPanel: PyDataViewerCommunityPanel
  private var jupyterSuggestionPanel: JPanel? = null

  init {
    isModal = false
    setCancelButtonText(PyBundle.message("debugger.data.view.close"))
    setCrossClosesWindow(true)

    dataViewerPanel = PyDataViewerCommunityPanel(PyDataViewerModel(myProject, value.frameAccessor))
    dataViewerPanel.apply(value, modifier = false, commandSource = null)
    dataViewerPanel.preferredSize = JBUI.size(TABLE_DEFAULT_WIDTH, TABLE_DEFAULT_HEIGHT)

    mainPanel = JPanel(BorderLayout())

    // If PythonPro installed and enabled.
    val jupyterSuggestionPanel = createJupyterSuggestionPanel()
    this.jupyterSuggestionPanel = jupyterSuggestionPanel
    if (jupyterSuggestionPanel != null) {
      mainPanel.add(jupyterSuggestionPanel, BorderLayout.NORTH)
    }

    mainPanel.add(dataViewerPanel, BorderLayout.CENTER)

    dataViewerPanel.addListener(object : PyDataViewerAbstractPanel.OnNameChangedListener {
      override fun onNameChanged(@NlsSafe name: @NlsContexts.TabTitle String) {
        title = name
      }
    })

    mainPanel.add(createBottomElements(), BorderLayout.SOUTH)

    title = value.fullName
    init()
  }

  override fun getDimensionServiceKey(): String = "#com.jetbrains.python.debugger.containerview.PyDataViewDialog"

  private fun createBottomElements(): JPanel {
    val colored = JBCheckBox(PyBundle.message("debugger.data.view.colored.cells"))
    colored.setSelected(isColoringEnabled(myProject))
    colored.addActionListener {
      setColoringEnabled(myProject, colored.isSelected)
      dataViewerPanel.dataViewerModel.isColored = colored.isSelected
    }

    val resize = JBCheckBox(PyBundle.message("debugger.data.view.resize.automatically"))
    resize.setSelected(isAutoResizeEnabled(myProject))
    resize.addActionListener {
      setAutoResizeEnabled(myProject, resize.isSelected)
      dataViewerPanel.resize(resize.isSelected)
      dataViewerPanel.updateUI()
    }

    return JPanel().apply {
      layout = BoxLayout(this, BoxLayout.Y_AXIS)
      add(colored)
      add(resize)
    }
  }

  override fun createActions(): Array<Action> = arrayOf(cancelAction)

  override fun createCenterPanel(): JPanel = mainPanel

  private fun createJupyterSuggestionPanel(): JPanel? {
    if (!isRichTableAndJupyterCanBeEnabled()) return null

    val requiredPluginIds = listOf(PluginId.getId("Pythonid"), PluginId.getId("intellij.jupyter"))

    var needToInstallPlugin = false
    val toInstallOrEnable = mutableSetOf<PluginId>().apply {
      requiredPluginIds.forEach { pluginId ->
        val pluginDescriptor = PluginManagerCore.findPlugin(pluginId)
        if (pluginDescriptor == null) {
          add(pluginId)
          needToInstallPlugin = true
        }
        else {
          if (!pluginDescriptor.isEnabled) {
            add(pluginId)
          }
        }
      }
    }

    if (toInstallOrEnable.isEmpty()) return null

    val panel = EditorNotificationPanel()
    panel.text = PyBundle.message("debugger.data.view.reach.view.suggestion")

    fun supposeRestartIfFRequired() {
      val requireRestart = toInstallOrEnable.firstOrNull { PluginManagerCore.getPlugin(it)?.isRequireRestart == true } != null
      if (requireRestart) {
        PluginManagerConfigurable.shutdownOrRestartApp()
      }
    }

    if (needToInstallPlugin) {
      panel.createActionLabel(IdeBundle.message("plugins.advertiser.action.install.plugin.name", "Jupyter")) {
        installAndEnable(myProject, toInstallOrEnable, showDialog = false) {
          hideJupyterSuggestionPanel()
          supposeRestartIfFRequired()
        }
      }
    }
    else {
      panel.createActionLabel(IdeBundle.message("link.enable.required.plugin", "Jupyter")) {
        toInstallOrEnable.forEach { PluginManagerCore.enablePlugin(it) }
        hideJupyterSuggestionPanel()
        supposeRestartIfFRequired()
      }
    }

    panel.createActionLabel(IdeBundle.message("plugins.advertiser.action.ignore.unknown.feature")) {
      disableJupyterSuggestions(myProject)
      hideJupyterSuggestionPanel()
    }

    return panel
  }

  /**
   * Determines if enhanced table features can be enabled.
   *
   * Checks three conditions:
   * 1. Running Ultimate Edition
   * 2. Jupyter suggestion is enabled
   * 3. Ultimate plugin is active
   */
  private fun isRichTableAndJupyterCanBeEnabled(): Boolean {
    val isUltimateEdition = !PlatformUtils.isCommunityEdition()
    val isJupyterEnabled = isJupyterSuggestionEnabled(myProject)
    val isUltimatePluginEnabled = !PluginManagerCore.isDisabled(PluginManagerCore.ULTIMATE_PLUGIN_ID)

    return isUltimateEdition
           && isJupyterEnabled
           && isUltimatePluginEnabled
  }

  private fun hideJupyterSuggestionPanel() {
    jupyterSuggestionPanel?.let {
      mainPanel.remove(it)
      mainPanel.revalidate()
      jupyterSuggestionPanel = null
    }
  }

  companion object {
    private const val TABLE_DEFAULT_WIDTH = 700
    private const val TABLE_DEFAULT_HEIGHT = 500

    private const val JUPYTER_SUGGESTION_ENABLED_PROPERTY_KEY = "python.debugger.dataView.jupyter.suggestion.enabled"

    private fun isJupyterSuggestionEnabled(project: Project) = PropertiesComponent.getInstance(project).getBoolean(JUPYTER_SUGGESTION_ENABLED_PROPERTY_KEY, true)

    private fun disableJupyterSuggestions(project: Project) = PropertiesComponent.getInstance(project).setValue(JUPYTER_SUGGESTION_ENABLED_PROPERTY_KEY, false, true)
  }
}